### ISSUE TITLE
Support for search / highlight

### ISSUE THREAD

**Issue by alexcherka** [kubetail-org/kubetail#13]
Would be really nice if you could filter console lines via search. Or even just highlight a desired key. I know this may not be the intention of this tool but when filtering back in time, would be great to search for specific data if needed!

**Comment by amorey** [comment]
Thanks for the suggestion! I definitely want to add search/highlight functionality to kubetail. It will take some time to implement but now's a good time to start specing it out. Is there another tool that you think has implemented log search particularly well so far?

**Comment by alexcherka** [comment]
As far as web k8s logs viewers go, I don't think there is much out there honestly... I know for IDE's I use K8s Lens that when you view a pod's logs it allows for searching via highlight and then jump around to results. 
![image](https://github.com/kubetail-org/kubetail/assets/9470827/de41dc63-a720-4261-b67f-f79b79b4a90e)

I think this is the best option depending on how you render your front end results. I assume there is some JS packages that allow for search/highlight content on page.

Found something that seems to cover it pretty well.
https://markjs.io/
Specifically the example [Search bar with jump to matches](https://jsfiddle.net/julmot/973gdh8g/)

**Comment by amorey** [comment]
Thanks! I'll take a look at K8S Lens and mark.js.

**Comment by nefarius** [comment]
I think in this case client-side search and indexing makes sense; no load increase on the backend and the client already has all the searchable data anyway so why not make use of it.

**Comment by alexcherka** [comment]
I agree with @nefarius, The front end data is loaded in the first place regardless and then filtering can be done client/DOM level and in theory no extra heavy lifting on the backend or really front end either. I think a highlighting approach is best and just have it jump around to matches. Similar to vscode in file search or any other ide as an example

**Comment by amorey** [comment]
Currently, the frontend only stores new events so the full log files aren't directly accessible locally. We could change this behavior and download the full files in the background and that would work well for small/medium files but it could be problematic for large amounts of data (and data egress costs could also be an issue down the road for power users). I'm still curious to test out client side search though because it might be useful for small amounts of data or in other contexts.

**Comment by sawan22071995** [comment]
Can you tell me how we will generated token when we use tokenauth.yaml. you didn't mention any commads or steps to get the token value to login in server.

**Comment by amorey** [comment]
Tokens are a Kubernetes authentication feature that you can use to add user authentication to your cluster. Depending on how you've configured your cluster they can be obtained from the cluster itself (Bootstrap Tokens) or from a 3rd-party authenticator plugin (e.g. aws-iam-authenticator). If you aren't using tokens to authenticate to your cluster you can use the `kubetail-clusterauth.yaml` manifest file to run Kubetail. If you'd like to start using tokens as part of your normal Kubernetes auth flow here are some places you can start:
https://kubernetes.io/docs/reference/access-authn-authz/authentication/
https://github.com/kubernetes-sigs/aws-iam-authenticator

I'm happy to answer any more questions you have about tokens but this Github issue is about search so please create a new issue if your question isn't search-related.

**Comment by amorey** [comment]
@alexcherka @nefarius @nodesocket @sawan22071995 

Hi folks!

I have an update about Kubetail search.

I just pushed a new `logs` command to the `kubetail` CLI tool that lets you grep your logs in real-time. It also lets you filter by time and source properties such as node and zone. To install the CLI tool you can download it from the [release page](https://github.com/kubetail-org/kubetail/releases/tag/cli%2Fv0.2.0) or using homebrew:

```console
brew install kubetail
```

Here are some examples of what you can do with the new `logs` command:

```console
# Tail 'web' deployment in the 'default' namespace
kubetail logs deployments/web

# Tail 'web' deployment in the 'frontend' namespace
kubetail logs frontend:deployments/web

# Return last 100 records
kubetail logs deployments/web --tail=100 

# Return first 100 records
kubetail logs deployments/web --head=100 

# Stream new records
kubetail logs deployments/web --follow

# Return all records
kubetail logs deployments/web --all

# Return first 10 records starting from 30 minutes ago
kubetail logs deployments/web --since PT30M

# Return last 10 records leading up to 30 minutes ago
kubetail logs deployments/web --until PT30M

# Return last 10 records that match "GET /about"
kubetail logs deployments/web --grep "GET /about" --force

# Return first 10 records that match "GET /about"
kubetail logs deployments/web --grep "GET /about" --head --force

# Return last 10 records that match "GET /about" or "GET /contact"
kubetail logs deployments/web --grep "GET /(about|contact)" --force

# Stream new records that match "GET /about"
kubetail logs deployments/web --grep "GET /about" --follow --force
```

Notice that to use the `--grep` filter you have to use the `--force` flag. This is because filtering is done client-side which means the tool will continuously download logs from your cluster until the desired number of matches are found which could result in unexpectedly large downloads. I'm working on a new feature to get around this issue.

Please check it out the new `logs` command and let me know what you think!

**Comment by amorey** [comment]
@alexcherka @nefarius @nodesocket @sawan22071995 

Hi everyone,

I'm happy to announce that search is live in the latest official release (cli/v0.4.3, helm/v0.10.1)! Try it out and let us know what you think!

Andres

**Comment by nefarius** [comment]
Hey hey @amorey I finally had the time to update to the latest version, looks and works great, thank you very much! 💪

**Comment by amorey** [comment]
Awesome! I'm very happy to hear you like it.
